dotnetv3/EC2/VirtualPrivateCloudExamples/CreateVPCforS3Example/CreateVPCforS3.cs

```diff
@@ -0,0 +1,136 @@
+using Amazon.S3;
+using Amazon.S3.Model;
+using Microsoft.Extensions.Configuration;
+using Tag = Amazon.EC2.Model.Tag;
+
+namespace CreateVPCforS3Example;
+
+/// <summary>
+/// Use a Amazon Elastic Compute Cloud (Amazon EC2) client to create a VPC Endpoint
+/// for the Amazon Simple Storage Service (Amazon S3) service and use
+/// that endpoint to list the objects in the S3 bucket.
+/// </summary>
+public class CreateVPCforS3
+{
+    static async Task Main(string[] args)
+    {
+        var ec2Client = new AmazonEC2Client();
+
+        var configuration = new ConfigurationBuilder()
+            .SetBasePath(Directory.GetCurrentDirectory())
+            .AddJsonFile("appsettings.json") // Load app settings from .json file.
+            .AddJsonFile("appsettings.local.json",
+                true) // Optionally, load local settings.
+            .Build();
+
+        var s3EndpointUrl = await CreateVPCforS3Client(configuration, ec2Client);
+
+        if (s3EndpointUrl != null)
+        {
+            var objectsList = await CreateS3ClientWithEndpoint(configuration, s3EndpointUrl);
+            Console.WriteLine("List of objects in bucket:");
+            Console.WriteLine(string.Join(" ", objectsList.Select(o => o.Key)));
+        }
+    }
+
+
+    // snippet-start:[EC2.dotnetv3.CreateVPCforS3]
+
+    /// <summary>
+    /// Create a VPC Endpoint and use it for an S3 client.
+    /// </summary>
+    /// <param name="configuration">Configuration to specify resource ids.</param>
+    /// <param name="ec2Client">Initialized EC2 client.</param>
+    /// <returns>The S3 url using the endpoint.</returns>
+    public static async Task<string?> CreateVPCforS3Client(IConfiguration configuration,
+        AmazonEC2Client ec2Client)
+    {
+        var endpointResponse = await ec2Client.CreateVpcEndpointAsync(
+            new CreateVpcEndpointRequest
+            {
+                VpcId = configuration["VpcId"],
+                VpcEndpointType = VpcEndpointType.Interface,
+                ServiceName = "com.amazonaws.us-east-1.s3",
+                SubnetIds = new List<string> { configuration["SubnetId"]! },
+                SecurityGroupIds = new List<string> { configuration["SecurityGroupId"]! },
+                TagSpecifications = new List<TagSpecification>
+                {
+                    new TagSpecification
+                    {
+                        ResourceType = ResourceType.VpcEndpoint,
+                        Tags = new List<Tag>
+                        {
+                            new Tag("service", "S3")
+                        }
+                    }
+                }
+            });
+
+        var newEndpoint = endpointResponse.VpcEndpoint;
+
+        Console.WriteLine(
+            $"VPC Endpoint {newEndpoint.VpcEndpointId} was created, waiting for it to be available. " +
+            $"This may take a few minutes.");
+
+        State? endpointState = null;
+
+        while (endpointState == null ||
```

Comment:
No waiter I presume?
Would this be a good candidate for extracting a function?

Reply:
Correct, no waiters in the .NET SDK. It's utility as a separate function would be limited outside of the creation operation, so I'd probably only move it out if we end up adding more to this example.